cloud/cloud_scheduler.cc

```diff
@@ -70,16 +71,25 @@ class LocalCloudScheduler : public CloudScheduler {
  public:
   LocalCloudScheduler(const std::shared_ptr<CloudScheduler>& scheduler,
                       long local_id)
-      : scheduler_(scheduler), next_local_id_(local_id) {}
+      : scheduler_(scheduler),
+        next_local_id_(local_id),
+        shutting_down_(false) {}
   ~LocalCloudScheduler() override {
-    for (const auto job : jobs_) {
+    {
+      std::lock_guard<std::mutex> lk(job_mutex_);
+      shutting_down_ = true;
+    }
+    for (const auto& job : jobs_) {
       scheduler_->CancelJob(job.second);
     }
     jobs_.clear();
   }
 
   long ScheduleJob(std::chrono::microseconds when,
                    std::function<void(void*)> callback, void* arg) override {
+    if (shutting_down_) {
```

Comment:
You're accessing a non-atomic boolean without a lock here. Is there some external synchronization to ensure this isn't racy?

How do you know the destructor has not completed? If it has, reading this boolean means reading undefined memory.